File: OsmAnd/src/net/osmand/plus/widgets/EmptyStateRecyclerView.java
package net.osmand.plus.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

public class EmptyStateRecyclerView extends RecyclerView {
	private View emptyView;

	public EmptyStateRecyclerView(Context context) {
		super(context);
	}

	public EmptyStateRecyclerView(Context context, @Nullable AttributeSet attrs) {
		super(context, attrs);
	}

	public EmptyStateRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
	}

	private final AdapterDataObserver emptyStateObserver = new AdapterDataObserver() {
		@Override
		public void onChanged() {
			checkIfEmpty();
		}

		@Override
		public void onItemRangeInserted(int positionStart, int itemCount) {
			checkIfEmpty();
		}

		@Override
		public void onItemRangeRemoved(int positionStart, int itemCount) {
			checkIfEmpty();
		}
	};

	@Override
	public void setAdapter(Adapter adapter) {
		Adapter oldAdapter = getAdapter();
		if (oldAdapter != null) {
			oldAdapter.unregisterAdapterDataObserver(emptyStateObserver);
		}
		super.setAdapter(adapter);
		if (adapter != null) {
			adapter.registerAdapterDataObserver(emptyStateObserver);
		}
		checkIfEmpty();
	}

	public void setEmptyView(View emptyView) {
		this.emptyView = emptyView;
		checkIfEmpty();
	}

	private void checkIfEmpty() {
		if (emptyView != null && getAdapter() != null) {
			if (getAdapter().getItemCount() == 0) {
				setVisibility(View.GONE);
				emptyView.setVisibility(View.VISIBLE);
			} else {
				emptyView.setVisibility(View.GONE);
				setVisibility(View.VISIBLE);
			}
		}
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/ui/views/EmptyStateRecyclerView.kt
package net.osmand.telegram.ui.views

import android.content.Context
import android.util.AttributeSet
import android.view.View

class EmptyStateRecyclerView : androidx.recyclerview.widget.RecyclerView {

	private var emptyView: View? = null

	private val emptyStateObserver = object : androidx.recyclerview.widget.RecyclerView.AdapterDataObserver() {
		override fun onChanged() {
			checkIfEmpty()
		}

		override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
			checkIfEmpty()
		}

		override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
			checkIfEmpty()
		}
	}

	constructor(context: Context) : super(context)

	constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

	constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
		context,
		attrs,
		defStyle
	)

	override fun setAdapter(adapter: androidx.recyclerview.widget.RecyclerView.Adapter<*>?) {
		val oldAdapter = getAdapter()
		oldAdapter?.unregisterAdapterDataObserver(emptyStateObserver)
		super.setAdapter(adapter)
		adapter?.registerAdapterDataObserver(emptyStateObserver)
		checkIfEmpty()
	}

	fun setEmptyView(emptyView: View) {
		this.emptyView = emptyView
		checkIfEmpty()
	}

	private fun checkIfEmpty() {
		adapter?.apply {
			val empty = itemCount == 0
			visibility = if (empty) View.GONE else View.VISIBLE
			emptyView?.visibility = if (empty) View.VISIBLE else View.GONE
		}
	}
}